blockchain/block/systemlog.go

```diff
@@ -0,0 +1,208 @@
+		if log := l.Proto(); log != nil {
+			sysLog.ImplicitTransferLog = append(sysLog.ImplicitTransferLog, log)
+			sysLog.NumTransactions++
+		}
+	}
+
+	if len(sysLog.ImplicitTransferLog) == 0 {
+		return nil
+	}
+	return &sysLog
+}
+
+// Proto returns the pb message
+func (log *ImplictTransferLog) Proto() *iotextypes.ImplicitTransferLog {
+	if len(log.txRecords) == 0 {
+		return nil
+	}
+
+	actionLog := iotextypes.ImplicitTransferLog{
+		ActionHash:   log.actHash[:],
+		Transactions: []*iotextypes.ImplicitTransferLog_Transaction{},
+	}
+	for _, l := range log.txRecords {
+		if record := l.toProto(); record != nil {
+			actionLog.Transactions = append(actionLog.Transactions, record)
+			actionLog.NumTransactions++
+		}
+	}
+
+	if actionLog.NumTransactions == 0 {
+		return nil
+	}
+	return &actionLog
+}
+
+func (log *TokenTxRecord) toProto() *iotextypes.ImplicitTransferLog_Transaction {
+	return &iotextypes.ImplicitTransferLog_Transaction{
+		Topic:     log.topic,
+		Amount:    log.amount,
+		Sender:    log.sender,
+		Recipient: log.recipient,
+	}
+}
+
+// SystemLogFromReceipt returns system logs in the receipt
+func SystemLogFromReceipt(receipts []*action.Receipt) *BlkImplictTransferLog {
+	if len(receipts) == 0 {
+		return nil
+	}
+
+	blkLog := BlkImplictTransferLog{
+		actionLogs: []*ImplictTransferLog{},
+	}
+	for _, r := range receipts {
+		if log := ReceiptSystemLog(r); log != nil {
+			blkLog.actionLogs = append(blkLog.actionLogs, log)
+			blkLog.numActions++
+		}
+	}
+
+	if blkLog.numActions == 0 {
+		return nil
+	}
+	return &blkLog
+}
+
+// ReceiptSystemLog generates system log from receipt
+func ReceiptSystemLog(r *action.Receipt) *ImplictTransferLog {
+	if r == nil || len(r.Logs) == 0 || r.Status != uint64(iotextypes.ReceiptStatus_Success) {
+		return nil
+	}
+
+	actionLog := ImplictTransferLog{
+		actHash:   r.ActionHash,
+		txRecords: []*TokenTxRecord{},
+	}
+	for _, log := range r.Logs {
+		if record := LogTokenTxRecord(log); record != nil {
+			actionLog.txRecords = append(actionLog.txRecords, record)
+			actionLog.numTxs++
+		}
+	}
+
+	if actionLog.numTxs == 0 {
+		return nil
+	}
+	return &actionLog
+}
+
+// LogTokenTxRecord generates token transaction record from log
+func LogTokenTxRecord(log *action.Log) *TokenTxRecord {
+	txRecord := TokenTxRecord{}
+
+	switch {
+	case log.IsEvmTransfer():
+		txRecord.topic = make([]byte, len(log.Topics[0]))
+		copy(txRecord.topic, log.Topics[0][:])
+		txRecord.amount = new(big.Int).SetBytes(log.Data).String()
+		from, _ := address.FromBytes(log.Topics[1][12:])
+		txRecord.sender = from.String()
+		to, _ := address.FromBytes(log.Topics[2][12:])
+		txRecord.recipient = to.String()
+		return &txRecord
+	case log.IsWithdrawBucket():
+		txRecord.topic = make([]byte, len(log.Topics[0]))
+		copy(txRecord.topic, log.Topics[0][:])
+		txRecord.amount = new(big.Int).SetBytes(log.Data).String()
+		txRecord.sender = log.Address
+		to, _ := address.FromBytes(log.Topics[2][12:])
```

Comment:
for stakewithdraw,sender is "staking",Recipient is the action caller,right?

Reply:
right, sender is staking protocol addr: io1qnpz47hx5q6r3w876axtrn6yz95d70cjl35r53